### PR TITLE
Allow pod.Spec.ActiveDeadlineSeconds to be updateable

### PR DESCRIPTION
**Pull request by pmorie**
Resurrection of #8890

@abhgupta @derekwaynecarr

### CODE REVIEW

**Comment by pmorie** [comment]
Still need to add E2Es

**Comment by derekwaynecarr** [comment]
I think you should be able to update the field as follows:

1. nil to X
2. X to Y where Y < X

any other update should fail validation.

This should not be used as a mechanism to pro-actively guarantee a pod could live longer than initially requested.

**Comment by pmorie** [comment]
@derekwaynecarr I'm down with that.  For E2E coverage, i'll run a pod that does `sleep 600`, and then drop the active deadline field down to 5 seconds and verify it gets killed.

**Comment by bgrant0607** [comment]
I'm fine with allowing updates to monotonically smaller values (including setting a deadline on a pod with no deadline).

**Comment by bgrant0607** [comment]
cc @kubernetes/goog-node

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -1380,6 +1380,26 @@ func ValidatePodUpdate(newPod, oldPod *api.Pod) field.ErrorList {
 		allErrs = append(allErrs, field.Forbidden(specPath.Child("containers"), "pod updates may not add or remove containers"))
 		return allErrs
 	}
+
+	// for updates, we are allowing ActiveDeadlineSeconds to be 0 since
+	// a user may just want to terminate the containers without deleting the pod
+	if newPod.Spec.ActiveDeadlineSeconds != nil {
+		newActiveDeadlineSeconds := *newPod.Spec.ActiveDeadlineSeconds
+		if newActiveDeadlineSeconds <= 0 {
+			allErrs = append(allErrs, field.Invalid(specPath.Child("activeDeadlineSeconds"), newActiveDeadlineSeconds, "activeDeadlineSeconds must be a positive integer greater than 0"))
```
Comment (derekwaynecarr) [inline pkg/api/validation/validation.go]: I think you can shorten the error message to just read:

"must be greater than 0"

The field path provides the context.

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -1380,6 +1380,26 @@ func ValidatePodUpdate(newPod, oldPod *api.Pod) field.ErrorList {
...
+			return allErrs
+		}
+		if oldPod.Spec.ActiveDeadlineSeconds != nil {
+			oldActiveDeadlineSeconds := *oldPod.Spec.ActiveDeadlineSeconds
+			if oldActiveDeadlineSeconds < newActiveDeadlineSeconds {
+				allErrs = append(allErrs, field.Invalid(specPath.Child("activeDeadlineSeconds"), newActiveDeadlineSeconds, "activeDeadlineSeconds may not update to a larger value"))
```
Comment (derekwaynecarr) [inline pkg/api/validation/validation.go]: "must be less than or equal to previous value"

File: pkg/api/validation/validation.go [hunk 3]
```diff
@@ -1380,6 +1380,26 @@ func ValidatePodUpdate(newPod, oldPod *api.Pod) field.ErrorList {
...
+				return allErrs
+			}
+		}
+	} else if oldPod.Spec.ActiveDeadlineSeconds != nil {
+		allErrs = append(allErrs, field.Invalid(specPath.Child("activeDeadlineSeconds"), newPod.Spec.ActiveDeadlineSeconds, "activeDeadlineSeconds may not be updated to a nil value"))
```
Comment (derekwaynecarr) [inline pkg/api/validation/validation.go]: must not update from a positive integer to nil value.

**Comment by pmorie** [comment]
Fixed @derekwaynecarr, will add e2e now that we agree on semantics

**Comment by pmorie** [comment]
@k8s-bot unit test this

**Comment by derekwaynecarr** [comment]
Food for future thought:
https://github.com/kubernetes/kubernetes/pull/19761#issuecomment-175841409

**Comment by pmorie** [comment]
@derekwaynecarr I think this is ready for review.

File: pkg/api/validation/validation.go [hunk 4]
```diff
@@ -1380,17 +1380,56 @@ func ValidatePodUpdate(newPod, oldPod *api.Pod) field.ErrorList {
...
+			allErrs = append(allErrs, field.Required(specPath.Child("containers").Index(i).Child("image"), ""))
+		}
+	}
+
+	// validate updated spec.activeDeadlineSeconds.  two types of updated are allowed:
+	// 1.  from nil to a positive value
+	// 2.  from a positive value to a lesser positive value
+	if newPod.Spec.ActiveDeadlineSeconds != nil {
+		newActiveDeadlineSeconds := *newPod.Spec.ActiveDeadlineSeconds
+		if newActiveDeadlineSeconds <= 0 {
```
Comment (liggitt) [inline pkg/api/validation/validation.go]: should we allow 0 here to mean "stop this now"? that seems a likely intent of someone decreasing the field value

Comment (pmorie) [inline pkg/api/validation/validation.go]: I think '1' effectively does the same thing, doesn't it?

Comment (liggitt) [inline pkg/api/validation/validation.go]: just a second, I'll let you know...

**Comment by derekwaynecarr** [comment]
I am now biased that we should not allow it to go from unset to set in an update.

File: pkg/api/validation/validation.go (hunk 4 above)
Comment (derekwaynecarr) [inline pkg/api/validation/validation.go]: I think @liggitt confirmed that we should accept 0.

**Comment by pmorie** [comment]
@derekwaynecarr @liggitt @abhgupta I'm now wondering if we need to support transitions _other_ than set -> zero

**Comment by pmorie** [comment]
@derekwaynecarr I understand your bias and share it due to the complication w/ quota and accounting

**Comment by pmorie** [comment]
I confirmed with abhgupta that the current OpenShift deployment code is just setting ActiveDeadlineSeconds to zero.  Seems like that should be the only transition we support to me, then.  @derekwaynecarr @bgrant0607 thoughts?

**Comment by bgrant0607** [comment]
@pmorie Why not delete with 0 grace period instead? Setting ActiveDeadlineSeconds to 0 seems like an obscure way to kill the pod.

**Comment by pmorie** [comment]
@bgrant0607 that might be a workable solution.  I am fairly certain that the commit we carry in OpenShift for this predates graceful deletion, which might be way we take this approach.  I'll talk with @abhgupta and @ironcladlou tomorrow to see if we really need this.

**Comment by bgrant0607** [comment]
@derekwaynecarr Why don't you want to support unset to set? How is that different from maxint64 to a smaller value?

**Comment by bgrant0607** [comment]
FWIW, we allow the deletion grace period to be (a) set and (b) reduced. It seems like we should be consistent with that.

**Comment by ironcladlou** [comment]
>@bgrant0607 that might be a workable solution. I am fairly certain that the commit we carry in OpenShift for this predates graceful deletion, which might be why we take this approach. I'll talk with @abhgupta and @ironcladlou tomorrow to see if we really need this.

We don't delete the pod because the user would lose access to the deployer pod's container/logs for post-cancellation inspection.

**Comment by bgrant0607** [comment]
Thanks, @ironcladlou. I figured it was something like that. We've discussed keeping pods around, such as indexed by uid, in several issues, such as #2789 and #14602, but haven't done so yet.

So, this seems like as a good a hack as any. :-)

**Comment by pmorie** [comment]
@bgrant0607 any opinion on whether we should allow you to update to a non-zero value?  Without a solid use-case for that, I'm tempted to only implement update-to-zero, and we can relax it later if someone has a legit use-case.

**Comment by bgrant0607** [comment]
Update to 0 only is harder to explain to users and different than the behavior of termination grace period. How hard is the implementation of more general updates?

**Comment by pmorie** [comment]
@bgrant0607 General updates are implemented in this PR currently.  Two types of update are allowed:

1.  Set `ActiveDeadlineSeconds` on an update when it is not already set
2.  Set `ActiveDeadlineSeconds` to a smaller value

If you're good with those, I think this PR should be ready for final review.  This PR also fixes a bug where it is possible to update a container with an empty image field.

**Comment by bgrant0607** [comment]
@pmorie Sounds fine to me.

File: pkg/api/validation/validation.go (hunk 4 above)
Comment (pmorie) [inline pkg/api/validation/validation.go]: I'll make it accept zero, I only realized now that I hadn't already done it.

**Comment by derekwaynecarr** [comment]
@pmorie - After more reflection, I am fine with my original proposal of supporting unset -> set, and update to a value less than the prior value.

File: pkg/api/validation/validation.go [hunk 5]
```diff
@@ -1380,17 +1380,56 @@ func ValidatePodUpdate(newPod, oldPod *api.Pod) field.ErrorList {
...
+	// 2.  spec.activeDeadlineSeconds
+
+	// validate updated container images
+	for i, ctr := range newPod.Spec.Containers {
+		if len(ctr.Image) == 0 {
+			allErrs = append(allErrs, field.Required(specPath.Child("containers").Index(i).Child("image"), ""))
+		}
+	}
+
+	// validate updated spec.activeDeadlineSeconds.  two types of updated are allowed:
```
Comment (derekwaynecarr) [inline pkg/api/validation/validation.go]: two types of updates are allowed

File: pkg/api/validation/validation.go [hunk 6]
```diff
@@ -1380,17 +1380,56 @@ func ValidatePodUpdate(newPod, oldPod *api.Pod) field.ErrorList {
...
+	// 1.  from nil to a positive value
+	// 2.  from a positive value to a lesser positive value
+	if newPod.Spec.ActiveDeadlineSeconds != nil {
+		newActiveDeadlineSeconds := *newPod.Spec.ActiveDeadlineSeconds
+		if newActiveDeadlineSeconds <= 0 {
+			allErrs = append(allErrs, field.Invalid(specPath.Child("activeDeadlineSeconds"), newActiveDeadlineSeconds, "must be greater than 0"))
```
Comment (derekwaynecarr) [inline pkg/api/validation/validation.go]: use `isNegativeErrorMsg`, and allow for 0

**Comment by pmorie** [comment]
@derekwaynecarr should be ready for final review, PTAL.  I will follow up w/ docs changes re: pod updates.

**Comment by pmorie** [comment]
@k8s-bot test this: #20694

**Comment by pmorie** [comment]
@k8s-bot test this issue: #20694

**Pull request merged**